File: Linguistics/Orthographies/OrthographyCounter.vb
''' <summary>
''' Implements a counter for glyphs of an orthography.
''' </summary>
Public NotInheritable Class OrthographyCounter
	Implements IReadOnlyDictionary(Of String, Int32)

	Private ReadOnly Glyphs As IDictionary(Of String, Int32)

	Friend Sub New(Orthography As Orthography)
		Glyphs = New Dictionary(Of String, Int32)(Orthography.Count)
		For Each Glyph In Orthography
			If Not Glyphs.ContainsKey(Glyph.ToUpper) Then
				Glyphs.Add(Glyph.ToUpper, 0)
			End If
		Next
	End Sub

	Public ReadOnly Property Count As Int32 Implements IReadOnlyCollection(Of KeyValuePair(Of String, Int32)).Count
		Get
			Return Glyphs.Count
		End Get
	End Property

	Public ReadOnly Property Keys As IEnumerable(Of String) Implements IReadOnlyDictionary(Of String, Int32).Keys
		Get
			Return Glyphs.Keys
		End Get
	End Property

	Public ReadOnly Property Values As IEnumerable(Of Int32) Implements IReadOnlyDictionary(Of String, Int32).Values
		Get
			Return Glyphs.Values
		End Get
	End Property

	Default Public ReadOnly Property Item(key As String) As Int32 Implements IReadOnlyDictionary(Of String, Int32).Item
		Get
			If ContainsKey(key) Then
				Return Glyphs(key.ToUpper)
			Else
				Return 0
			End If
		End Get
	End Property

	Public Sub Add(Source As String)
		For Each S In Source.ToUpper
			If ContainsKey(S) Then
				Glyphs(S) += 1
			End If
		Next
	End Sub

	Public Sub Clear()
		For i = 0 To Count - 1
			Glyphs(i) = 0
		Next
	End Sub

	Public Function ContainsKey(key As String) As Boolean Implements IReadOnlyDictionary(Of String, Int32).ContainsKey
		Return Glyphs.ContainsKey(key.ToUpper)
	End Function

	Public Function GetEnumerator() As IEnumerator(Of KeyValuePair(Of String, Int32)) Implements IEnumerable(Of KeyValuePair(Of String, Int32)).GetEnumerator
		Return Glyphs.GetEnumerator()
	End Function

	Public Function TryGetValue(key As String, ByRef value As Int32) As Boolean Implements IReadOnlyDictionary(Of String, Int32).TryGetValue
		Return Glyphs.TryGetValue(key.ToUpper, value)
	End Function

	Private Function IEnumerable_GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
		Return GetEnumerator()
	End Function
End Class

File: Literary/OrthographyCounter.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace Stringier.Literary {
	/// <summary>
	/// Implements a counter for glyphs of an orthography.
	/// </summary>
	internal sealed class OrthographyCounter : IReadOnlyDictionary<Glyph, Int32> {
		private readonly IDictionary<Glyph, Int32> Glyphs;

		internal OrthographyCounter(Orthography orthography) {
			Glyphs = new Dictionary<Glyph, Int32>(orthography.Count);
			foreach (Glyph glyph in orthography) {
				if (!Glyphs.ContainsKey(glyph.ToUpper())) {
					Glyphs.Add(glyph.ToUpper(), 0);
				}
			}
		}

		/// <inheritdoc/>
		public Int32 this[Glyph key] {
			get {
				if (ContainsKey(key)) {
					return Glyphs[key.ToUpper()];
				} else {
					return 0;
				}
			}
		}

		public Int32 this[Char key] => this[new Glyph(key)];

		/// <inheritdoc/>
		public IEnumerable<Glyph> Keys => Glyphs.Keys;

		/// <inheritdoc/>
		public IEnumerable<Int32> Values => Glyphs.Values;

		/// <inheritdoc/>
		public Int32 Count {
			get {
				Int32 count = 0;
				foreach (Int32 value in Values) {
					count += value;
				}
				return count;
			}
		}

		public void Add(String @string) {
			foreach (Glyph glyph in @string.EnumerateGlyphs()) {
				if (ContainsKey(glyph)) {
					Glyphs[glyph.ToUpper()]++;
				}
			}
		}

		/// <inheritdoc/>
		public Boolean ContainsKey(Glyph key) => Glyphs.ContainsKey(key.ToUpper());

		public Boolean ContainsKey(Char key) => ContainsKey(new Glyph(key));

		/// <inheritdoc/>
		public IEnumerator<KeyValuePair<Glyph, Int32>> GetEnumerator() => Glyphs.GetEnumerator();

		/// <inheritdoc/>
		public Boolean TryGetValue(Glyph key, out Int32 value) => Glyphs.TryGetValue(key.ToUpper(), out value);

		public Boolean TryGetValue(Char key, out Int32 value) => TryGetValue(new Glyph(key), out value);

		/// <inheritdoc/>
		IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
	}
}
